lib/Reader/readerir.cpp

```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
+}
+
+void GenIR::popDebugScope() {
+  LLILCDbgInfo.LexicalBlocks.pop_back();
+}
+
+// Set the Debug Location for the current instruction
+void GenIR::setDebugLocation(uint32_t CurrOffset, uint32_t IsCall) {
+  DIScope *Scope;
+  if (LLILCDbgInfo.LexicalBlocks.empty())
+    Scope = LLILCDbgInfo.TheCU;
+  else
+    Scope = LLILCDbgInfo.LexicalBlocks.back();
+  DebugLoc Loc = DebugLoc::get(CurrOffset, IsCall, Scope);
+
+  LLVMBuilder->SetCurrentDebugLocation(Loc);
+}
+
+llvm::DISubroutineType *GenIR::createFunctionType(llvm::Function *F,
+                                                  DIFile *Unit) {
+  SmallVector<Metadata *, 8> EltTys;
+  // for each param, etc. convert the type to DIType and add it to the array.
+  FunctionType *FunctionTy = F->getFunctionType();
+
+  unsigned i = 0;
+
+  DIType *ReturnTy = convertType(Function->getReturnType());
+  EltTys.push_back(ReturnTy);
+
+  for (auto ParamIterator = FunctionTy->param_begin();
+       ParamIterator != FunctionTy->param_end(); ParamIterator++) {
+    Type *Ty = *ParamIterator;
+    DIType *ParamTy = convertType(Ty);
+
+    EltTys.push_back(ParamTy);
+    i++;
+  }
+  return DBuilder->createSubroutineType(Unit,
+                                        DBuilder->getOrCreateTypeArray(EltTys));
+}
+
+llvm::DIType *GenIR::convertType(Type *Ty) {
+  StringRef TyName;
+  llvm::dwarf::TypeKind Encoding;
+
+  if (Ty->isVoidTy()) {
+    return nullptr;
+  }
+  if (Ty->isIntegerTy()) {
+    Encoding = llvm::dwarf::DW_ATE_signed;
+    unsigned BitWidth = Ty->getIntegerBitWidth();
+
+    switch (BitWidth) {
+    case 8:
+      TyName = "char";
+      Encoding = llvm::dwarf::DW_ATE_unsigned_char;
+      break;
+    case 16:
+      TyName = "short";
+      break;
+    case 32:
+      TyName = "int";
+      break;
+    case 64:
+      TyName = "long int";
+      break;
+    case 128:
+      TyName = "long long int";
+      break;
+    default:
+      TyName = "int";
+      break;
+    }
+    llvm::DIType *DbgTy =
+        DBuilder->createBasicType(TyName, BitWidth, BitWidth, Encoding);
+    return DbgTy;
+  }
+  if (Ty->isFloatingPointTy()) {
+    Encoding = llvm::dwarf::DW_ATE_float;
+    unsigned BitWidth = Ty->getPrimitiveSizeInBits();
+    if (Ty->isHalfTy()) {
+      TyName = "half";
+    }
+    if (Ty->isFloatTy()) {
+      TyName = "float";
+    }
+    if (Ty->isDoubleTy()) {
+      TyName = "double";
+    }
+    if (Ty->isX86_FP80Ty()) {
+      TyName = "long double";
+    }
+
+    llvm::DIType *DbgTy =
+        DBuilder->createBasicType(TyName, BitWidth, BitWidth, Encoding);
+    return DbgTy;
+  }
+
+  if (Ty->isPointerTy()) {
+    uint64_t Size = JitContext->TM->getDataLayout()->getPointerTypeSize(Ty);
+    uint64_t Align = JitContext->TM->getDataLayout()->getPrefTypeAlignment(Ty);
+    llvm::DIType *DbgTy = DBuilder->createPointerType(
+        convertType(Ty->getContainedType(0)), Size, Align);
+
+    return DbgTy;
+  }
+  // TODO: add support for aggregate types
+  return nullptr;
```

Comment:
Is it OK to just return `nullptr` from this function? Do we understand the impact of doing so? I don't think that we should see FCAs in our function signatures at the moment in any case, so it might be better to just turn this into an `assert(UNREACHED)` or `llvm_unreachable()`.